pydantic/_internal/_model_construction.py

```diff
@@ -553,6 +556,59 @@ def complete_model_class(
     return True
 
 
+def set_deprecated_descriptors(cls: type[BaseModel]) -> None:
+    """Set data descriptors on the class for deprecated fields."""
+    for field, field_info in cls.model_fields.items():
+        if (msg := field_info.deprecation_message) is not None:
+            desc = _DeprecatedFieldDescriptor(msg)
+            desc.__set_name__(cls, field)
+            setattr(cls, field, desc)
+
+    for field, computed_field_info in cls.model_computed_fields.items():
+        if (
+            not computed_field_info.from_deprecated_decorator  # Avoid having two warnings emitted
+            and (msg := computed_field_info.deprecation_message) is not None
+        ):
+            desc = _DeprecatedFieldDescriptor(msg, computed_field_info.wrapped_property)
+            desc.__set_name__(cls, field)
+            setattr(cls, field, desc)
+
+
+class _DeprecatedFieldDescriptor:
```

Comment:
Nice, I see we could also add more support for stacklevel, etc stuff here in the future.

Reply:
I like this approach in general, but want to check what @dmontagu thinks. My main concern here is that we might be overlooking something important in terms of setting and getting model attributes with this custom descriptor, so I want to make sure this isn't a footgun in that sense.

Reply:
I think this is unlikely to cause issues except with deprecated fields, and if it does, we can worry about it then.

I will be surprised if this doesn't introduce some theoretical misbehavior, but I also won't be surprised if no one ever hits it (and if they do, I strongly suspect there will be easy workarounds until such misbehavior gets fixed).

As long as there are some tests of setting attributes with this thing, I think it should be fine.

The main thing I would probably try to make sure still behaves correctly is that it does the right thing if `model_config['validate_assignment'] = True`.